search/search_query.cpp

```diff
@@ -1609,44 +1615,99 @@ void Query::SearchLocality(MwmValue const * pMwm, Locality & res1, Region & res2
 
   auto codingParams = trie::GetCodingParams(pMwm->GetHeader().GetDefCodingParams());
 
-  ModelReaderPtr searchReader = pMwm->m_cont.GetReader(SEARCH_INDEX_FILE_TAG);
-
-  using TValue = FeatureIndexValue;
-  auto const trieRoot = trie::ReadTrie<SubReaderWrapper<Reader>, ValueList<TValue>>(
-      SubReaderWrapper<Reader>(searchReader.GetPtr()), SingleValueSerializer<TValue>(codingParams));
-
-  ForEachLangPrefix(params, *trieRoot, [&](TrieRootPrefix & langRoot, int8_t lang)
+  auto versionFormat = pMwm->GetHeader().GetFormat();
+  if (versionFormat < version::v7)
   {
-    impl::DoFindLocality doFind(*this, pMwm, lang);
-    MatchTokensInTrie(params.m_tokens, langRoot, doFind);
+    using TValue = FeatureWithRankAndCenter;
+
+    ModelReaderPtr searchReader = pMwm->m_cont.GetReader(SEARCH_INDEX_FILE_TAG);
+
+    auto const trieRoot = trie::ReadTrie<SubReaderWrapper<Reader>, ValueList<TValue>>(
+        SubReaderWrapper<Reader>(searchReader.GetPtr()),
+        SingleValueSerializer<TValue>(codingParams));
+
+    ForEachLangPrefix(params, *trieRoot, [&](TrieRootPrefix<TValue> & langRoot, int8_t lang)
+                      {
+                        impl::DoFindLocality doFind(*this, pMwm, lang);
+                        MatchTokensInTrie(params.m_tokens, langRoot, doFind);
+
+                        // Last token's prefix is used as a complete token here, to limit number of
+                        // results.
+                        doFind.Resize(params.m_tokens.size() + 1);
+                        doFind.SwitchTo(params.m_tokens.size());
+                        MatchTokenInTrie(params.m_prefixTokens, langRoot, doFind);
+                        doFind.SortLocalities();
+
+                        // Get regions from STATE and COUNTRY localities
+                        vector<Region> regions;
+                        doFind.GetRegions(regions);
+
+                        // Get best CITY locality.
+                        Locality loc;
+                        doFind.GetBestCity(loc, regions);
+                        if (res1 < loc)
+                        {
+                          LOG(LDEBUG, ("Better location ", loc, " for language ", lang));
+                          res1.Swap(loc);
+                        }
+
+                        // Get best region.
+                        if (!regions.empty())
+                        {
+                          sort(regions.begin(), regions.end());
+                          if (res2 < regions.back())
+                            res2.Swap(regions.back());
+                        }
+                      });
+  }
+  else if (versionFormat == version::v7)
```

Comment:
Please, create a temporary class in search, like MwmFeatures:
```c++
class MwmFeatures {
public:
  enum class SearchIndexFormat {
    FeaturesWithRankAndCenter,
    CompressedBitVector,
    Unknown
  };

  MwmFeatures(versionFormat);

  SearchIndexFormat GetSearchIndexFormat() const;  
};
```
And use it instead of these comparisons. Later we will move it to indexer and re-use in other places.